nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -1377,41 +1362,80 @@ impl DataStore {
+                    return Err(Error::invalid_request(
                         "cannot delete instance: instance is running or has \
                                 not yet fully stopped",
-                    );
-                    }
-                    let instance_state =
-                        collection.runtime_state.nexus_state.state();
-                    match instance_state {
-                        api::external::InstanceState::Stopped
-                        | api::external::InstanceState::Failed => {
-                            Error::internal_error("cannot delete instance")
-                        }
-                        _ => Error::invalid_request(&format!(
-                            "instance cannot be deleted in state \"{}\"",
-                            instance_state,
-                        )),
-                    }
+                    ));
                 }
-                DetachManyError::DatabaseError(e) => {
-                    public_error_from_diesel(e, ErrorHandler::Server)
+                let instance_state =
+                    collection.runtime_state.nexus_state.state();
+                match instance_state {
+                    api::external::InstanceState::Stopped
+                    | api::external::InstanceState::Failed => {
+                        Err(Error::internal_error("cannot delete instance"))
+                    }
+                    _ => Err(Error::invalid_request(&format!(
+                        "instance cannot be deleted in state \"{}\"",
+                        instance_state,
+                    ))),
                 }
-            })?;
+            }
+            DetachManyError::DatabaseError(e) => {
+                Err(public_error_from_diesel(e, ErrorHandler::Server))
+            }
+        })?;
+
+        Ok(())
+    }
 
+    /// Delete the provided `authz_instance`, as long as it is in one of the
+    /// provided set of [`InstanceState`]s.
+    ///
+    /// A.K.A. "[`project_delete_instance`], hard mode". Typically, instances
+    /// may only be deleted if they are in the [`InstanceState::NoVmm`] or
+    /// `[`InstanceState::Failed`] states. This method exists separately in
+    /// order to allow an unwinding `instance-create` saga to delete the
+    /// instance record without transiently moving it to
+    /// [`InstanceState::Failed`], in which it becomes eligible for automatic
+    /// restart.
+    ///
+    /// In general, callers should use [`project_delete_instance`] instead,
+    /// unless you really know what you're doing.
+    ///
+    /// [`project_delete_instance`]: Self::project_delete_instance
+    pub async fn project_delete_instance_in_states(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        ok_to_delete_instance_states: &'static [InstanceState],
+    ) -> DeleteResult {
+        // First, mark the instance record as deleted
```

Comment:
similarly to my comment on naming, maybe we should state that this also detaches disks etc? i might also want to see this comment explicitly state that it's the part that changes the instance's state to `Destroyed`, and maybe also mention the idempotency here?

Reply:
Sure, I'll elaborate on this. I'll also clarify why we're doing it first